src/ray/common/client_connection.cc

```diff
@@ -78,11 +81,75 @@ ray::Status ServerConnection<T>::WriteMessage(int64_t type, int64_t length,
   message_buffers.push_back(boost::asio::buffer(&type, sizeof(type)));
   message_buffers.push_back(boost::asio::buffer(&length, sizeof(length)));
   message_buffers.push_back(boost::asio::buffer(message, length));
-  // Write the message and then wait for more messages.
-  // TODO(swang): Does this need to be an async write?
   return WriteBuffer(message_buffers);
 }
 
+template <class T>
+void ServerConnection<T>::WriteMessageAsync(
+    int64_t type, int64_t length, const uint8_t *message,
+    const std::function<void(const ray::Status &)> &handler) {
+  auto write_buffer = std::unique_ptr<AsyncWriteBuffer>(new AsyncWriteBuffer());
+  write_buffer->write_version = RayConfig::instance().ray_protocol_version();
+  write_buffer->write_type = type;
+  write_buffer->write_length = length;
+  write_buffer->write_message.resize(length);
+  write_buffer->write_message.assign(message, message + length);
+  write_buffer->handler = handler;
+
+  // TODO(ekl) should we enforce a max number of async bytes outstanding here?
```

Comment:
Maybe log a warning every time exceeds the next power of 2? E.g., print warnings when it crosses 2**20, 2**21, ... What do you think?

Reply:
I added a warn, let's see if its' useful.